### ISSUE TITLE
Improve documentation on how to write an EF Core provider 

### ISSUE THREAD

**Issue by divega** [dotnet/EntityFramework.Docs#681]
Currently we have something very brief at https://docs.microsoft.com/ef/core/providers/writing-a-provider pointing to @ajcvickers's blog.  

It would be good to update this to cover:
- Support for EF Core 2.1 features
- The mechanics of setting up spec tests (e.g. https://github.com/dotnet/efcore/issues/27107#issuecomment-1004923695)
- The mechanics of adding provider specific extension methods

Whether this should be in a blog post or if it needs to be part of the documentation is something we can discuss, but I would be inclined to have it in the actual documentation.

**Comment by TomasHubelbauer** [comment]
I'd like to express my preference for including the guide on how to create one's own EF Core database provider in the documentation as well. It's way more authoritative than a blog post. I'm just trying to figure out going by the old version of the docs, can you provide a rundown of the changes or a super rough estimate for when 3.1.0 might land?

**Comment by omatrot** [comment]
I'd like to write a scaffolding only provider around edmx files in order to fill the gap for all of us having existing edmx files that we want to keep around in order to create entities and DbContexts.

Any information that could allow me to start would be very much appreciated.

**Comment by ajcvickers** [comment]
Suggestion from @ErikEJ on https://github.com/dotnet/efcore/issues/20793#issuecomment-621943308

> It would lower the barrier to have a "template/null" provider to use as a starting point. But of course it can quickly become stale.

**Comment by roji** [comment]
Ideally this would be built as part of CI, to keep it in at least minimal sync with EF...

**Comment by ErikEJ** [comment]
@roji do you think there would be any value in this?

**Comment by roji** [comment]
Yeah, I think it would... Our docs repo has a set of samples - which now even get built in CI - so it could live there. On the other hand, if we put it in the normal repo it would naturally evolve as we change EF Core...

**Comment by ErikEJ** [comment]
I think it is better to fix it to depend on a released version of EF Core, and then just update on new release (if needed at all)

**Comment by smitpatel** [comment]
Releases to consider
- LTS release
- Current release (stable)
- Current preview

I think current preview would provide slightly lesser value. If I am writing a provider from scratch, I would want to depend on stable release so I can build my provider and release it. It would be easy after 3.1 to move to next major release for providers even with breaking changes. The issue with targeting previews would be constantly moving target. If EF Core added some API in one preview but changed it in another preview (we love API reviews), then provider has unnecessary effort to keep changing things when writing the core provider at the same time.

**Comment by ErikEJ** [comment]
@smitpatel exactly my experience/pain during EF 7 previews

**Comment by roji** [comment]
I think we should be much more stable now than we used to be - but I agree that there's not that much value to a provider template that tracks previews. So maybe just tracking the current stable release is enough (an additional one for LTS seems to be a bit overkill)? If so, then just putting it in our doc repo under samples should be good...

On another note, I assume we're discussing about a *relational* provider template - or do we think that a non-relational sample is valuable as well?

Are you interested in doing this @ErikEJ?

**Comment by ErikEJ** [comment]
@roji yes, the relational. I could just base it on something similar to the initial commit I mentioned above. A no-name ANSI SQL provider...

**Comment by ajcvickers** [comment]
If we put it in the main repo, then it would evolve along with the other providers _and_ there will still always be a branch/tag to get the snapshot of the provider that aligns with a given release.

**Comment by ErikEJ** [comment]
Good point, but should the initial commit not target 3.1?

**Comment by ajcvickers** [comment]
Yeah, that would be ideal.

**Comment by roji** [comment]
@ajcvickers the point made above is that the main value of this is for the latest stable release, since that's what provider writers are likely to target. If that is so, then the template would depend on the latest released version and not on the latest code in master... and we'd evolve it for every release manually, a bit like how we're supposed to do for the doc samples.

**Comment by smitpatel** [comment]
The repo to host is less important. The sample should use package references rather than project references.

**Comment by ErikEJ** [comment]
@smitpatel good point - just let me know which repo you prefer 😊

**Comment by roji** [comment]
If we agree to use package references, I'd vote for putting it in the docs repo. It really is a sample, and has no value being in the main repo.

**Comment by ajcvickers** [comment]
@ErikEJ We discussed this and we're thinking that we could try this out on https://github.com/efcore first to see what it looks like and how it evolves. We could potentially move it somewhere else later.

**Comment by ErikEJ** [comment]
Yes, I will discuss details in the newly created repo there...

**Comment by dellams** [comment]
Are there any plans to have the docs updated for EFCore5?

The blog it points to is for EFCore1.1! Is the doc still valid? I cannot find DatabaseProviderServices?

I searched in the linked example providers such as SQL Server, npsql, etc but could not find DatabaseProviderServices or RelationalDatabaseProviderServices anywhere in them...

Any ideas? I really need to write a provider but haven't a clue how to or where to start?

Any help would be really appreciated.

Many thanks,
David.

**Comment by roji** [comment]
@dellams we're indeed lacking good documentation on how to write a provider... Assuming you're looking to develop a provider for a relational database, it's recommended you start with the [Npgsql](https://github.com/npgsql/efcore.pg) or [Pomelo MySQL](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql) providers, as these are external projects that aren't part of the EF Core repo. Where relevant, the other providers (Sqlite, SQL Server) can also be consulted.

Start by looking at the different services implemented by these providers, and getting familiar with the general architecture of EF Core. We can then help out with any specific questions.

**Comment by dellams** [comment]
Yes, thank you, I have already checked them out and was surprised with just how much work and how complex it was to write a provider. I was hoping you could provide a simple bare bones template with the minimum needed to implement a provider?

Do you know when some up to date documentation will be created for this?

What I am trying to do is something a bit more special than just a relational provider, I am attempting to integrate EFCore into the OASIS API I have been building, which integrates all blockchains, holochain, cloud, db's, networks, platforms, APIs, etc. It is like a network of networks, an abstraction layer over the entire internet.

I would love to get some support with this very valuable project, I feel once people understood what I am building a LOT would come and help me build it! :)

https://github.com/NextGenSoftwareUK/Our-World-OASIS-API-HoloNET-HoloUnity-And-.NET-HDK

It would be super cool and powerful to have this abstraction API using EFCore making it super easy to bridge everything to everything through one simple ORM (I love EF and have worked with it many years being a big fan of .NET since its initial release many years ago just after I graduated from University in 2002.) :)

I have worked as a IT consultant/contractor in London building commercial .NET applications of all sizes in all markets, etc so have experience of most of the amazing .NET stack since I have grown as a developer along with .NET... you could say we grew up together! ;-)

BIG FAN! :) Keep up the great work, so happy it is Open Source now and cross platform,... very excited and passionate to help grow the .NET ecosystem with you guys...

I hope to get on the .NET Foundation soon... is that something you guys could help with? When you all understand the wonders I plan to do for the .NET community I think you would want to help me help you... ;-)

Many thanks,
David.

**Comment by ErikEJ** [comment]
@dellams you may find the series of commits in the develop branch here useful: https://github.com/GibraltarSoftware/VistaDB.EFCore

**Comment by dellams** [comment]
Great thank you. Is this an example of the steps bones minimum needed? 🙂

**Comment by ErikEJ** [comment]
It depends...

**Comment by ElanHasson** [comment]
@dellams sounds like you're describing [F# Type Providers](https://docs.microsoft.com/en-us/dotnet/fsharp/tutorials/type-providers/)

**Comment by mguinness** [comment]
One of the providers that is stalled out is [FileContextCore](https://github.com/morrisjdev/FileContextCore).  The fact that there are so many breaking changes between major versions means that ongoing maintenance of providers is very labor intensive especially when there is no documentation.

**Comment by DaveRMaltby** [comment]
I have a need to create many EF Core providers and decided to create a template repo for this purpose.  It is still a work in progress (i.e. need to get passing unit tests going), but am wondering if [this repo](https://github.com/Servant-Software-LLC/EFCore.SampleProvider) would be something that the community would be interested in?  And also if it might go a ways toward solving this open issue?

The [readme](https://github.com/Servant-Software-LLC/EFCore.SampleProvider#readme) describes the concept of this sample provider.

**Comment by roji** [comment]
@DaveRMaltby I do think we skeleton/template provider is a good thing to have, as a starting point for people writing providers.

Some specific points come to mind when looking at your repo; in particular, many services are being overridden that would only be needed in relatively advanced scenarios (e.g. SqlExpressionFactory only for custom SQL expressions, QueryTranslationPostprocessorFactory only for certain advanced scenarios, etc.). I'd advise going with a more minimum approach, implementing only services which a typical provider would have to override, rather than everything.

**Comment by DaveRMaltby** [comment]
> @DaveRMaltby I do think we skeleton/template provider is a good thing to have, as a starting point for people writing providers.
> 
> Some specific points come to mind when looking at your repo; in particular, many services are being overridden that would only be needed in relatively advanced scenarios (e.g. SqlExpressionFactory only for custom SQL expressions, QueryTranslationPostprocessorFactory only for certain advanced scenarios, etc.). I'd advise going with a more minimum approach, implementing only services which a typical provider would have to override, rather than everything.

@roji, thank you for the feedback.  Yes, your advice points to the issue that I'm struggling with and I'm sure others who have no idea about the nature of the services offered here.  I just used the [Npgsql](https://github.com/npgsql/efcore.pg), [Pomelo MySQL](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql) and  [SQLite](https://github.com/dotnet/efcore/tree/main/src/EFCore.Sqlite.Core) providers as a basis to copy from for this [template repo](https://github.com/Servant-Software-LLC/EFCore.SampleProvider) without understanding their purposes.  I was trying to get at least one of the tests in EFCore.Relational.Specification.Tests to pass.  So far, I have been unsucessful in that endevour and it has been overwhelming as just simply adding the EFCore.SampleProvider.FunctionalTests.**Query** tests amounts to 443 tests.

Anyhow, I believe now that I need to take a completely different approach.    I've now stepped back and am going to take one of the many EF Core providers that I plan to write and create a few unit tests based off of the examples in [Getting Started with EF Core](https://learn.microsoft.com/en-us/ef/core/get-started/overview/first-app?tabs=netcore-cli).  Getting those CRUD tests to pass will give me the understanding and basis for the bare minimum services that are needed.  Later, I will add more advanced features to this provider as needed (eventually, adding the migrations/scaffolding that I also need).  Ultimately, I'll try to circle around and update the SampleProvider to provide a skeleton/template for others to make use. (And also I hope to provide descriptions about the services as well)

Can you possibly provide a quick list of services that you know are required just to achieve the first step mentioned above of getting simple CRUD unit tests to pass?  Thank you.

**Comment by ErikEJ** [comment]
@DaveRMaltby Maybe some of the initial commits here could help? https://github.com/GibraltarSoftware/VistaDB.EFCore/compare/main...develop

**Comment by roji** [comment]
@DaveRMaltby it's unfortunate that we don't have better docs for provider writers - it really is something we want to improve, but tends to be de-prioritized in favor of user-facing features. However, we're here to help you out with any questions - please don't hesitate to ask.

**Comment by DaveRMaltby** [comment]
> @DaveRMaltby it's unfortunate that we don't have better docs for provider writers - it really is something we want to improve, but tends to be de-prioritized in favor of user-facing features. However, we're here to help you out with any questions - please don't hesitate to ask.

@ErikEJ and @roji, other priorities took me away from this topic.  I'm trying to focus again here on providing a useful [template repo](https://github.com/Servant-Software-LLC/EFCore.SampleProvider) for the community.  I don't like the direction that I was initially going there.  I'm now thinking that such a template repo should just be able to take any full fledged [ADO.NET data provider](https://learn.microsoft.com/en-us/dotnet/framework/data/adonet/data-providers) (that supports .NET Core, of course) as it basis without needing much else.  Is this a naive statement?  I will likely try to go in this direction, as it may allow most of the specific database logic for anyone who needs to create an EF Core provider to be placed in their own ADO.NET data provider and then use this template repo to get most of the way there quickly.

**Comment by mguinness** [comment]
You could also take a look at [FileBaseContext](https://github.com/dualbios/FileBaseContext) which is probably one of the simpler providers out there.  Maybe @dualbios can provide some tips on creating a provider from scratch.

There's also preliminary work being done on a provider for LiteDB at https://github.com/mbdavid/LiteDB/issues/1198 which might be of interest.

**Comment by roji** [comment]
@DaveRMaltby I'm not quite sure what you mean with "may allow most of the specific database logic for anyone who needs to create an EF Core provider to be placed in their own ADO.NET data provider and then use this template repo to get most of the way there very quickly"; there's typically various database-specific logic that has to go into each EF provider, which does not belong in the lower-level ADO.NET provider (e.g. how to produce various SQL constructs). The idea of "general ADO.NET-compatible EF provider" which would somehow just work with any ADO.NET provider simply doesn't work.

I think the main value in a template here would be to have a minimal provider sample which implements the services that **must** be overridden (or would typically be overridden for most providers). This would only provide a starting point for new providers, or for understanding how a provider looks like.

Note also that a minimal relational (ADO.NET-based) EF provider looks very different from a non-relational provider; for example, a relational EF provider typically needs to have its version of QuerySqlGenerator, which non-relational provider obviously don't need. I agree it's worth concentrating on a relational template; non-relational providers in general can be extremely different based on the database being targeted, and there's not likely to be a lot of commonality there.

**Comment by DaveRMaltby** [comment]
@roji and @ErikEJ, I didn't forget about this.  I have finally completed what I was referring to above.  I have taken and built a custom EF Core provider that depends on a SQLite ADO.NET provider.  Of course, I knew that there would be some custom code that is unique to support the SQLite provider, but it doesn't seem that much as you can see in [MyCustom EF Core provider repo](https://github.com/Servant-Software-LLC/EFCore.MyCustom).  
See what yáll think.  Look at the [README](https://github.com/Servant-Software-LLC/EFCore.MyCustom/blob/main/README.md) for this repo.  I've tried to logically break up what is needed into 4 steps.  The basic unit tests are passing for me.
Note:  It builds against EF Core 6.x, but I only did that because currently my product is using that version.  If/when you want to add this into an EF Core repo, of course we'll rev it up to the latest version.
Anyhow, let me know if you want more of my involvement here.  For my company's use, I have 4 providers to implement and will be working on those efforts, unless you need something further here.  Thanks for the guidance!

**Comment by JohnGoldInc** [comment]
So is it a correct statement that the critical part is working outwards from the ...ServiceCollectionExtensions classes' "IServiceCollection AddEntityFramework..." section, and create your versions of those classes listed there as needed? And that InMemoryServiceCollectionExtensions.cs or CosmosServiceCollectionExtensions.cs is the version probably close to the minimum needed to implement?

**Comment by DaveRMaltby** [comment]
> So is it a correct statement that the critical part is working outwards from the ...ServiceCollectionExtensions classes' "IServiceCollection AddEntityFramework..." section, and create your versions of those classes listed there as needed? And that InMemoryServiceCollectionExtensions.cs or CosmosServiceCollectionExtensions.cs is the version probably close to the minimum needed to implement?

Honestly, I'm still not an expert and got a lot of guidance from ChatGPT.  I asked it to indicate the minimal classes required.  Anyhow, the ServiceCollectionExtensions class does register your provider services that the Entity Framework will use via DI for itself.   It is only your DbContext derived classes ([BloggingContext](https://github.com/Servant-Software-LLC/EFCore.MyCustom/blob/main/EFCore.MyCustom.Tests/Models/BloggingContext.cs) in my unit tests) that end up kicking off the framework to do things using your provider registered services.  Anyhow, this was the minimal classes that I found was need to get the [GettingStartedTests](https://github.com/Servant-Software-LLC/EFCore.MyCustom/blob/main/EFCore.MyCustom.Tests/GettingStartedTests.cs) running with the SQLite in-memory ADO.NET provider.  Note:  This hasn't been reviewed by @roji and @ErikEJ as far as know, so this example may be missing things.

**Comment by ThaDaVos** [comment]
So any idea when the docs get updated?

**Comment by roji** [comment]
I'd highly recommend using AI to explore things rather than waiting for documentation on this topic.

**Comment by ThaDaVos** [comment]
Maybe you should use AI to fix the docs then...

**Comment by anasik** [comment]
I wrote a successor to @ajcvickers original post that's featured in the docs and I would like it to be featured the same way.

Here it is: https://anasismail.com/so-you-still-want-to-write-an-ef-core-provider/

Let me know if I am allowed to make a PR with this?

@AndriySvyryd

**Comment by AndriySvyryd** [comment]
@anasik Sure
